Prototype: single-namespace

caseydavenport opened the pull request.

caseydavenport committed:
Prototype ability for operator to install multiple copies of
tigera-manager

caseydavenport commented on pkg/controller/certificatemanager/certificatemanager.go:
```diff
@@ -85,14 +85,18 @@ type CertificateManager interface {
 // Create creates a signer of new certificates and has methods to retrieve existing KeyPairs and Certificates. If a user
 // brings their own secrets, CertificateManager will preserve and return them.
-func Create(cli client.Client, installation *operatorv1.InstallationSpec, clusterDomain string) (CertificateManager, error) {
+func Create(cli client.Client, installation *operatorv1.InstallationSpec, clusterDomain, ns string) (CertificateManager, error) {
```
This function now takes a namespace.

For single-tenant clusters, this namespace will be passes a "tigera-operator"

For multi-tenant clusters, this will be the tenant's namespace.

caseydavenport commented on pkg/controller/manager/manager_controller.go:
```diff
@@ -209,13 +217,18 @@ type ReconcileManager struct {
 
 // GetManager returns the default manager instance with defaults populated.
-func GetManager(ctx context.Context, cli client.Client) (*operatorv1.Manager, error) {
+func GetManager(ctx context.Context, cli client.Client, ns string) (*operatorv1.Manager, error) {
```
This now takes the namespace to query the Manager instance from.

caseydavenport commented on pkg/controller/manager/manager_controller.go:
```diff
@@ -264,183 +289,208 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
-		render.ProjectCalicoAPIServerTLSSecretName(installation.Variant),
-		render.TigeraLinseedSecret,
+	trustedSecretNames := []render.Secret{
+		render.PacketCaptureServerCert(r.multiTenant, request.InstallNamespace()),
```
Trying out an approach here. The secrets that we fetch may be in one of two namespaces:

- 1: single-tenant, will be in `tigera-operator`
- 2: multi-tenant, will be in the tenant's namespace (which varies for each call to Reconcile())

So, this approach turns the constant into a function, that takes the install namespace and returns a struct instead of a string.... (We can actually remove the first argument from this function which makes it simpler).

caseydavenport commented on pkg/controller/manager/manager_controller.go:
```diff
@@ -264,183 +289,208 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
+	certificateManager.AddToStatusManager(r.status, request.InstallNamespace())
 
 	// Check that Prometheus is running
+	// TODO: We'll need to run an instance of Prometheus per-tenant? Or do we use labels to delimit metrics?
```
We actually don't need to run Prometheus in the management cluster except for our own metrics. We don't store product data in the management cluster prometheus AFAIK

caseydavenport commented on pkg/controller/manager/manager_controller.go:
```diff
@@ -264,183 +289,208 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
 
 	// Get secrets used by the manager to authenticate with Elasticsearch.
+	// TODO: esSecrets should be namespaced in the tenant's namespace. We can't get rid of this 100%
+	// since some secrets are needed for Kibana auth flow.
```
For management clusters that don't have Kibana enabled, we can skip over this entirely.

caseydavenport commented on pkg/controller/manager/manager_controller.go:
```diff
@@ -540,46 +592,56 @@ func (r *ReconcileManager) Reconcile(ctx context.Context, request reconcile.Requ
+				rcertificatemanagement.NewKeyPairOption(certificateManager.KeyPair(), true, false),
 				rcertificatemanagement.NewKeyPairOption(tlsSecret, true, true),
 				rcertificatemanagement.NewKeyPairOption(linseedVoltronSecret, true, true),
-				rcertificatemanagement.NewKeyPairOption(internalTrafficSecret, false, true),
```
Whoops, this was an accidental change. Can revert this.

caseydavenport commented on pkg/controller/manager/request.go:
```diff
@@ -0,0 +1,49 @@
+// TruthNamespace returns the namespace to use as the source of truth for storing data.
+// For single-tenant installs, this is the tigera-operator namespace.
+// For multi-tenant installs, this is tenant's namespace.
+func (r *Request) TruthNamespace() string {
```
Might be a better name for this, but this is what I came up with for now.

The idea is that there are two logical namespaces to consider:

- The "source of truth" namespace. This is where secrets and config are installed and persisted. In traditional operator installs, this is `tigera-operator`. In multi-tenant installs, this would be the tenant's own namespace instead.
- The namespace where components are installed. In traditional operator installs, this would be "tigera-manager". In multi-tenant installs this would be the tenant's namespace as well.

So, for multi-tenant installs these two namespaces are the same. But for single-tenant installs they would be different.

caseydavenport commented on pkg/controller/utils/utils.go:
```diff
@@ -180,9 +182,6 @@ func WaitToAddTierWatch(tierName string, controller controller.Controller, c kub
 // will be generated.
 func AddNamespacedWatch(c controller.Controller, obj client.Object, metaMatches ...MetaMatch) error {
 	objMeta := obj.(metav1.ObjectMetaAccessor).GetObjectMeta()
-	if objMeta.GetNamespace() == "" {
```
Pretty sure this was just over-aggressive validation - the code after this seems to support this being empty correctly.

caseydavenport commented on pkg/render/certificatemanagement/certificatemanagement.go:
```diff
@@ -51,30 +75,38 @@ func (c component) ResolveImages(*operatorv1.ImageSet) error {
-				if keyPairCreator.renderInNamespace {
-					objsToCreate = append(objsToCreate, keyPair.Secret(c.cfg.Namespace))
-				}
+		if keyPair == nil {
```
This is just a rephrasing of the previous way of doing this that results in less indentation.

caseydavenport commented on pkg/render/guardian.go:
```diff
@@ -127,9 +127,12 @@ func (c *GuardianComponent) Objects() ([]client.Object, []client.Object) {
 		// Add tigera-manager service account for impersonation
 		CreateNamespace(ManagerNamespace, c.cfg.Installation.KubernetesProvider, PSSRestricted),
-		managerServiceAccount(),
+		managerServiceAccount("tigera-manager"), // TODO
```
This actually can probably be left as it is. Voltron will need to translate serviceaccount names.

caseydavenport commented on pkg/render/manager.go:
```diff
@@ -198,16 +203,18 @@ func (c *managerComponent) SupportedOSType() rmeta.OSType {
 	objs := []client.Object{
-		CreateNamespace(ManagerNamespace, c.cfg.Installation.KubernetesProvider, PSSRestricted),
+		// TODO: The namespace should be pre-created in multi-tenant environments.
+		// Maybe we should do this for single-tenant as well, to keep them more similar?
```
It's actually not a problem to pre-create and then create it again here.

We'll need to decide whether we keep the owner reference though - it means that creation / deletion of the namespace isn't symmetric if we do. But that might not be a problem.

caseydavenport commented on pkg/render/manager.go:
```diff
@@ -153,7 +157,8 @@ type ManagerConfiguration struct {
 	// Whether the cluster supports pod security policies.
-	UsePSP bool
+	UsePSP    bool
+	Namespace string
```
We need to pass the namespace in to the render component now instead of hard-coding it.

Would be good to _force_ this behavior - e.g., delete the `render.ManagerNamespace` variable. This will ensure every component is doing the right thing.

caseydavenport commented on pkg/render/manager.go:
```diff
@@ -374,8 +381,10 @@ func (c *managerComponent) managerProxyProbe() *corev1.Probe {
 // managerEnvVars returns the envvars for the manager container.
 func (c *managerComponent) managerEnvVars() []corev1.EnvVar {
 	envs := []corev1.EnvVar{
+		// TODO: Prometheus URL will need to change.
```
Maybe - Prometheus might be OK actually.

rene-dekker commented on pkg/controller/certificatemanager/certificatemanager.go:
```diff
@@ -110,17 +114,17 @@ func Create(cli client.Client, installation *operatorv1.InstallationSpec, cluste
-		}, caSecret)
-		if err != nil && !kerrors.IsNotFound(err) {
+		k := types.NamespacedName{Name: certificatemanagement.CASecretName, Namespace: ns}
+		if err = cli.Get(context.Background(), k, caSecret); err != nil && !kerrors.IsNotFound(err) {
```
We want a different CA for each tenant's namespace? Why is that?

caseydavenport replied:
The advantage being that each tenant's components will be issued certificates from their own CA - invalidating a single tenant becomes trivial (just delete the CA and regenerate it)

rene-dekker replied:
I'm not convinced I see a use-case for that. Aren't we over-complicating things? The tenant's components are still just our components inside of our cluster, they just happen to serve a particular tenant. 

I'd argue if we want isolation of these tenant namespaces, we should be using network policies.

caseydavenport replied:
> I'd argue if we want isolation of these tenant namespaces, we should be using network policies.

We'll do both. I agree that it's not a requirement - a single CA for all services would work as well, but we need a separate CA for each tenant's Voltron mTLS tunnel anyway, so using that CA for other services seems like an easy defense-in-depth improvement.

rene-dekker commented on pkg/controller/request.go:
```diff
@@ -0,0 +1,75 @@
+// For multi-tenant installs, this is tenant's namespace.
+func (r *CommonRequest) TruthNamespace() string {
+	if !r.multiTenant {
+		return "tigera-operator"
```
Should use common.OperatorNamespace in case you run the operator elsewhere.

rene-dekker commented on pkg/controller/utils/auth.go:
```diff
@@ -75,6 +75,7 @@ func GetKeyValidatorConfig(ctx context.Context, cli client.Client, authenticatio
 
 // GetIdpSecret retrieves the Secret containing sensitive information for the configuration IdP specified in the given
 // operatorv1.Authentication CR.
+// TODO: This needs to be updated to support multiple tenants IDP secrets.
```
The authentication spec will likely not be MT. Users log in through the cloud portal (Auth0). The idp secret is then the secret for the cloud portal in this case, and so it won't be MT either in that case.

rene-dekker commented on pkg/render/tenant.go:
```diff
@@ -0,0 +1,69 @@
+
+var _ Component = &tenantIsolater{}
+
+func NewTenantIsolater(c Component, ns string) Component {
```
This func is not used anywhere?

caseydavenport replied:
Oh yeah that's from an earlier prototype, will remove.

nelljerram committed:
Set DNS trusted servers using FelixConfig instead of FELIX_ env var

For Rancher and OpenShift the operator needs to set a non-default value for DNS trusted servers, and
it was previously doing that by setting a FELIX_ env var in the node daemonset.  But that makes it
impossible for any further tweaks to that using FelixConfiguration - because FelixConfiguration
settings are always overridden by FELIX_ env vars.

To allow later tweaks, change the operator to create or patch FelixConfiguration instead.

References:
https://tigera.atlassian.net/browse/RS-897
https://tigera.atlassian.net/browse/EV-3193

nelljerram committed:
make generate

nelljerram committed:
Don't keep adding same DNS server, and don't change FelixConfig when not needed

rene-dekker committed:
Fix missing certificate management configurations (#2697)

* "Fix missing certificate management configurations"

* Add tests for certificate management fixes.

* "Add extra certificate management tests and fixes."

caseydavenport committed:
Linseed must be able to verify Voltron's certificate

caseydavenport committed:
Merge pull request #2706 from tigera/casey-fix-linseed-voltron

Linseed must be able to verify Voltron's certificate

asincu committed:
Drop cluster name from intrusion detection controller (#2710)

nelljerram committed:
Merge remote-tracking branch 'origin/master' into dns-trusted-servers

nelljerram committed:
Merge pull request #2700 from tigera/dns-trusted-servers

Set DNS trusted servers using FelixConfig instead of FELIX_ env var

caseydavenport committed:
Revert LS changes
